### ISSUE TITLE
How to disable show errors in other files?

### ISSUE THREAD

**Issue by AllenSue** [flycheck/flycheck#1434]
This new feature is kind of bother me because my project has some code is correct, but flychecker backend can't parse them, and report errors, flycheck shows many errors I want to ignore from other file.

**Comment by cpitclaudel** [comment]
It's not super easy to disable the feature ATM, but we can definitely add an option :) Can you help by clarifying which language and which checker you're using, and ideally giving a small example?

**Comment by cpitclaudel** [comment]
You could do this in the meantime, I think:

```
(advice-add flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil))
```

**Comment by cpitclaudel** [comment]
CC @purcell

**Comment by AllenSue** [comment]
I'm writting C, and use `cquery`, checker is `lsp-ui-flycheck`, someone defined a macro in my project:
```C
#define container_of(ptr, type, member) ({			\
        const typeof( ((type *)0)->member ) *__mptr = (ptr);	\
        (type *)( (char *)__mptr - offsetof(type,member) );})
```
and used many times, looks like:
```C
    struct user_structure {
        int member;
    };
    
    foo(int* var) {
        struct user_structure* us = container_of(var, struct user_structure, member);
    }
```
and will report errors:
![flycheck](https://user-images.githubusercontent.com/20827968/37503309-be851ac2-2912-11e8-8d4d-a12d7aab43d9.png)

**Comment by AllenSue** [comment]
I'm using `spacemacs`, and I put code below in `dotspacemacs/user-config()`:
```elisp
  (with-eval-after-load 'flycheck
    (advice-add flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)) 
  )
```
Emacs will report error when startup:
```
(Spacemacs) Error in dotspacemacs/user-config: Symbol’s value as variable is void: flycheck-relevant-error-other-file-p
```
Am I wrong?

**Comment by cpitclaudel** [comment]
Woops:

    (with-eval-after-load 'flycheck
      (advice-add 'flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)) 

I was missing a `'`

**Comment by AllenSue** [comment]
OK, it works, thanks!

**Comment by fmdkdd** [comment]
Suggesting that users disable our functions with an advice in their config seems a bad idea in the long run.

I would be in favor of adding a toggle.  Just checking the toggle in `flycheck-relevant-error-other-file-p` would have the same effect as the advice, right?

**Comment by cpitclaudel** [comment]
> Suggesting that users disable our functions with an advice in their config seems a bad idea in the long run.

Yes, of course :) I hope it was clear I wasn't recommending this in the long run.

> Just checking the toggle in flycheck-relevant-error-other-file-p would have the same effect as the advice, right?

I think so, though the check would likely be best put in flycheck-relevant-error-p.

**Comment by AllenSue** [comment]
That sounds good. @purcell 

I'm not familiar with flycheck or flycheckers, I just use them. But should the certain checkers do something to adapt this feature to? I only write C, maybe there are differences among the different language.

I have a simple example:
![example](https://user-images.githubusercontent.com/20827968/37562123-e193c9da-2a9a-11e8-8bf7-f2ac5a42f9c2.png)
`hello.c` doesn't depend on `main.c`, but still can see error in it. I think this does not match the intention you said above.

**Comment by alexmurray** [comment]
I also see the same thing @AllenSue  - using only cquery with lsp-ui for flycheck - so I think this is really a bug in cquery / emacs-cquery - perhaps @MaskRay / @topisani / @sebastiencs can help?

Basically the issue is now that flycheck shows errors which occur in other files (which I guess is to indicate say an error in a header file which gets included in some other file) - when using cquery / lsp-ui flycheck is listing error messages for files which aren't even related to the current one - see hello.c and main.c above in https://github.com/flycheck/flycheck/issues/1434#issuecomment-373969238

So it seems cquery / lsp-ui / lsp-mode is listing errors which are not relevant to the current buffer file.

**Comment by alexmurray** [comment]
Yeah it looks like lsp-ui does flycheck for the whole workspace - https://github.com/emacs-lsp/lsp-ui/blob/master/lsp-ui-flycheck.el - note all the references to 'the current workspace' - so then the question is should lsp-ui be changed to only show errors for the current file?

**Comment by alexmurray** [comment]
Opened an issue with lsp-ui https://github.com/emacs-lsp/lsp-ui/issues/98 since I think that is the correct place to fix this (not in flycheck itself - although a buffer-local defcustom for this behaviour would be useful in the meantime anyway)

**Comment by cpitclaudel** [comment]
Thanks all for investigating and taking steps to improve the matter :)
I expect there will be cases where the problem cannot be easily fixed upstream, (think sphinx-build), and for these cases I think we'll likely add an `:error-filter`, likely predicated on a defcustom (something like `flycheck-show-errors-in-other-files` with values `all`, `related`, and `none`.)

**Comment by cpitclaudel** [comment]
> I now really think that a single global setting would be the wrong way to go: 

Yup, I agree with you

> So I suspect that generating a defvar per checker would be the way to go, in the same way that the -executable vars are produced. 

Not sure; see below.

>  Or it might be that we should have to explicitly add to a checker definitions an :other-file-errors flag in order for other-file errors to be accepted from that checker.

I don't think we need a new property — just a change in the error filter.  I think it's OK display other errors by default, and add a filter where they aren't relevant,

And that suggests a good way to let user customize this, too: we can add a `flycheck-silence-other-file-errors` function which uses `add-function` to wrap the existing checker filter.

**Comment by purcell** [comment]
> I don't think we need a new property — just a change in the error filter. I think it's OK display other errors by default, and add a filter where they aren't relevant,

Ah, that makes sense!

**Comment by NimoTh** [comment]
> Woops:
> 
> (with-eval-after-load 'flycheck
>   (advice-add 'flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)) 
> I was missing a '

That one was still missing a closing parenthesis ). This one works for me (although for some reason, he still looks inside two .h includes. He does skip .hpp includes.

```
(with-eval-after-load 'flycheck
  (advice-add 'flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)))
```

**Comment by NimoTh** [comment]
> Woops:
> 
> ```
> (with-eval-after-load 'flycheck
>   (advice-add 'flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)) 
> 
> ```
> I was missing a '

That was still missing a closing parenthesis ). This one works great for me (although errors for .h includes still seem show up for some reason, .hpp is ignored):

```
(with-eval-after-load 'flycheck
  (advice-add 'flycheck-relevant-error-other-file-p :override (lambda (&rest args) nil)))
```

**Comment by cpitclaudel** [comment]
We now have `flycheck-relevant-error-other-file-show`, so I think we're good on this issue.
